subsys/sip_svc/sip_svc.c

```diff
@@ -0,0 +1,898 @@
+
+		if (!ctrl->tid) {
+			LOG_ERR("Thread not spawned during init");
+			sip_svc_id_map_remove_item(ctrl->trans_id_map, trans_id);
+			sip_svc_id_mgr_free(ctrl->clients[c_idx].trans_idx_pool,
+						trans_idx);
+			k_mutex_unlock(&ctrl->data_mutex);
+			return SIP_SVC_ID_INVALID;
+		}
+		k_thread_resume(ctrl->tid);
+		k_mutex_unlock(&ctrl->data_mutex);
+	}
+	return trans_id;
+}
+
+void *sip_svc_get_priv_data(struct sip_svc_controller *ctrl,
+			       uint32_t c_token)
+{
+	uint32_t c_idx;
+
+	if (!ctrl)
+		return NULL;
+
+	if (k_mutex_lock(&ctrl->data_mutex, K_FOREVER) == 0) {
+		c_idx = sip_svc_get_c_idx(ctrl, c_token);
+
+		if (c_idx == SIP_SVC_ID_INVALID) {
+			k_mutex_unlock(&ctrl->data_mutex);
+			return NULL;
+		}
+
+		k_mutex_unlock(&ctrl->data_mutex);
+		return ctrl->clients[c_idx].priv_data;
+	}
+
+	return NULL;
+}
+
+struct sip_svc_controller *sip_svc_get_controller(char *method)
+{
+	struct sip_svc_controller *cnt = controller;
+
+	while (strlen(cnt->method) != 0) {
+		if (!strcmp(cnt->method, method)) {
+			return cnt;
+		}
+		cnt++;
+	}
+	return NULL;
+}
+
+static int sip_svc_init(const struct device *unused)
+{
+	ARG_UNUSED(unused);
+	struct sip_svc_controller *ctrl = NULL;
+	struct sip_svc_client *client = NULL;
+	size_t msgq_size = sizeof(struct sip_svc_request) *
+				CONFIG_ARM_SIP_SVC_MSGQ_DEPTH;
+	char *msgq_buf = NULL;
+	uint32_t i;
+	int ret = 0;
+
+	/* Get controller array */
+	for (ctrl = &controller[0] ; ctrl->invoke_fn != NULL ; ctrl++) {
+		/* Return if controller has been initialized */
+		if (ctrl->init)
+			continue;
+
+		ctrl->client_id_pool = sip_svc_id_mgr_create(
+						CONFIG_ARM_SIP_SVC_MAX_CLIENT_COUNT);
+		if (!ctrl->client_id_pool)
+			return -ENOMEM;
+
+
+		ctrl->trans_id_map = sip_svc_id_map_create(
+					CONFIG_ARM_SIP_SVC_MSGQ_DEPTH);
+		if (!ctrl->trans_id_map) {
+			sip_svc_id_mgr_delete(ctrl->client_id_pool);
+			return -ENOMEM;
+		}
+
+		/* Alloc request msgq ring buffer */
+		msgq_buf = k_malloc(msgq_size);
+		if (!msgq_buf) {
+			sip_svc_id_mgr_delete(ctrl->client_id_pool);
+			sip_svc_id_map_delete(ctrl->trans_id_map);
+			return -ENOMEM;
+		}
+
+		/* Initialize request msgq */
+		k_msgq_init(&ctrl->req_msgq, msgq_buf,
+				sizeof(struct sip_svc_request),
+				CONFIG_ARM_SIP_SVC_MSGQ_DEPTH);
+
+		/* Initialize request msgq mutex */
+		k_mutex_init(&ctrl->req_msgq_mutex);
+
+		/* Reset thread pointer */
+		ctrl->tid = NULL;
+
+		/* Initialize client contents */
+		for (i = 0; i < CONFIG_ARM_SIP_SVC_MAX_CLIENT_COUNT; i++) {
+			client = &ctrl->clients[i];
+			client->id = SIP_SVC_ID_INVALID;
+			client->token = SIP_SVC_ID_INVALID;
+			client->state = SIP_SVC_CLIENT_ST_INVALID;
+			client->active_trans_cnt = 0;
+
+			client->trans_idx_pool = sip_svc_id_mgr_create(
+						CONFIG_ARM_SIP_SVC_MAX_TRANSACTION_COUNT);
+			if (!client->trans_idx_pool) {
+				ret = -ENOMEM;
+				break;
+			}
+		}
+
+		if (ret != 0) {
+			sip_svc_id_mgr_delete(ctrl->client_id_pool);
+			k_free(ctrl->trans_id_map);
```

Comment:
line 834 you are using `sip_svc_id_map_delete(ctrl->trans_id_map);` but here is `k_free(ctrl->trans_id_map);` ?

Reply:
fixed